pkg/controllers/cluster/taint_manager.go

```diff
@@ -0,0 +1,360 @@
+		tc.clusterBindingEvictionWorker.Add(key)
+	}
+	return controllerruntime.Result{RequeueAfter: tc.ClusterTaintEvictionRetryFrequency}, nil
+}
+
+// Start starts an asynchronous loop that handle evictions.
+func (tc *NoExecuteTaintManager) Start(ctx context.Context) error {
+	bindingEvictionWorkerOptions := util.Options{
+		Name:          "binding-eviction",
+		KeyFunc:       nil,
+		ReconcileFunc: tc.syncBindingEviction,
+	}
+	tc.bindingEvictionWorker = util.NewAsyncWorker(bindingEvictionWorkerOptions)
+	tc.bindingEvictionWorker.Run(tc.ConcurrentReconciles, ctx.Done())
+
+	clusterBindingEvictionWorkerOptions := util.Options{
+		Name:          "cluster-binding-eviction",
+		KeyFunc:       nil,
+		ReconcileFunc: tc.syncClusterBindingEviction,
+	}
+	tc.clusterBindingEvictionWorker = util.NewAsyncWorker(clusterBindingEvictionWorkerOptions)
+	tc.clusterBindingEvictionWorker.Run(tc.ConcurrentReconciles, ctx.Done())
+
+	<-ctx.Done()
+	return nil
+}
+
+func (tc *NoExecuteTaintManager) syncBindingEviction(key util.QueueKey) error {
+	fedKey, ok := key.(keys.FederatedKey)
+	if !ok {
+		klog.Errorf("Failed to sync binding eviction as invalid key: %v", key)
+		return fmt.Errorf("invalid key")
+	}
+	cluster := fedKey.Cluster
+	keyStr := fedKey.String()
+
+	binding := &workv1alpha2.ResourceBinding{}
+	if err := tc.Client.Get(context.TODO(), types.NamespacedName{Namespace: fedKey.Namespace, Name: fedKey.Name}, binding); err != nil {
+		// The resource no longer exist, in which case we stop processing.
+		if apierrors.IsNotFound(err) {
+			tc.evictionStartTimeMap.delete(keyStr)
+			return nil
+		}
+		return fmt.Errorf("failed to get binding %s: %v", fedKey.NamespaceKey(), err)
+	}
+
+	if !binding.DeletionTimestamp.IsZero() || !util.IsTargetClusterExist(binding.Spec.Clusters, cluster) {
+		tc.evictionStartTimeMap.delete(keyStr)
+		return nil
+	}
+
+	needEviction, tolerationTime, err := tc.needEviction(cluster, binding.Annotations)
+	if err != nil {
+		klog.ErrorS(err, "Failed to check if binding needs eviction", "binding", klog.KObj(binding))
+		return err
+	}
+
+	// Case 1: Need eviction now.
+	// Case 2: need eviction after toleration time. If time is up, do eviction right now.
+	// Case 3: Tolerate forever, we do nothing.
+	evictNow := false
+	if needEviction || tolerationTime == 0 {
+		evictNow = true
+	} else if tolerationTime > 0 {
+		now := time.Now()
+		if startTime, exist := tc.evictionStartTimeMap.get(keyStr); exist {
+			// If trigger time has not been reached, just return.
+			if startTime.Add(tolerationTime).After(now) {
+				return nil
+			}
+			// If triggered, evict now.
+			evictNow = true
+		} else {
+			// If first, we set the start time and add this object after tolerationTime.
+			tc.evictionStartTimeMap.set(keyStr, now)
+			tc.bindingEvictionWorker.AddAfter(fedKey, tolerationTime)
+			return nil
+		}
+	}
+
+	if evictNow {
+		if result := util.RemoveTargetCluster(binding.Spec.Clusters, cluster); len(result) < len(binding.Spec.Clusters) {
```

Comment:
Note: This is the thing I want to optimize. 
I hope the `evicted` cluster could be exist in somewhere until the scheduler finds a new cluster for the resource referencing by rb/crb.

I'll start working on it.

Reply:
cc @lonelyCZ who might be interested in this feature I'm talking about.
(Sorry, for the delay, I'm afraid I don't have enough time for it this week, I'll try to get back on this next week.)

Reply:
OK, Thanks for your hard work.

Reply:
Another question, I guess the `len(result) < len(binding.Spec.Clusters)` want to confirm if the cluster is still in the `target`, if yes, can we use something like:
```golang
		if binding.Spec.TargetContains(cluster) {
			binding.Spec.Clusters = util.RemoveTargetCluster(binding.Spec.Clusters, cluster)
			...
		}
```
would it looks better?

Reply:
I have already checked here first, see L141.

```go
	if !binding.DeletionTimestamp.IsZero() || !binding.Spec.TargetContains(cluster) {
		return nil
	}
```

So `len(result) < len(binding.Spec.Clusters)` is just a double check, no meaning here. I could remove this check.

Reply:
Yeah, I agree to remove the redundant check.